### DISCUSSION TITLE
"Error: No matching key" logged in the wallet and preventing any WC action with connected dapps

### DISCUSSION THREAD

**Discussion by jeanregisser** [WalletConnect/walletconnect-monorepo#3291] [Developer Support] [answered]
Hey,

We're struggling with an error happening from time to time. We've updated to the latest release of `@walletconnect` libs (2.9.1) and are still seeing it:

```
{ context: 'client' }, [Error: No matching key. session: 9a05b6099c6cc662ed39dd43795adec83ad64dfc6cfd7779ae1bc3d6470bb5c6]
```

This is happening in [Valora](https://github.com/valora-inc/wallet), in a React Native environment (0.70.12).
We don't yet have reproduction steps. But we've seen it when testing with https://curve.fi so far.

The symptom is the user does an action in a dapp that is already connected to Valora and Valora doesn't show the modal to approve it. But instead logs the error above.

Connecting to new dapps still work after that, but any action from the dapp doesn't show the modal in Valora and logs the error.

We've seen that clearing local storage allows fixing the issue. But for real users, the only option to recover from this is to reinstall Valora and re-import their account. Which is a big problem.

For our users, it looks like Valora is broken because it doesn't show the modal to approve the action.

We've searched through the WalletConnect repo and found many related discussions:
https://github.com/search?q=org%3AWalletConnect+Error%3A+%22No+matching+key.%22&type=discussions

But none so far giving a definitive way to fix it for good.

Do you have any recommendation on what to do next?
Let us know if you need more information.

**Comment by MuckT** [comment]
@glitch-txs would you be able to take a look at this? 🙏

**Comment by ganchoradkov** [comment]
Hey @jeanregisser, `no matching key` is a general store error that has different variants and causes depending on the type, in your case, it can't find a `session` with a given topic. Are there additional logs preceding the error that might hint at what might be going on?

Are you doing some custom cleanup/management of sessions with the client in your wallet?
Generally, being subscribed to `session_delete` so you can get notified when a session is disconnected by the peer, should be sufficient to keep your store synced.

**Reply by jeanregisser** [reply]
Thanks for looking into this @ganchoradkov 🙌 

We're already subscribed to `session_delete` ([here](https://github.com/valora-inc/wallet/blob/927c5ba67c3263ba2c79abb57bc75612cccf5464/src/walletConnect/saga.ts#L158-L160)) and aren't doing anything fancy with sessions.
Though we store some data about the sessions in our Redux store, we always go back to the WC `web3wallet` client instance to act on them. The complete implementation is in [src/walletConnect/saga.ts](https://github.com/valora-inc/wallet/blob/8c7bdd04b682c305a3c533c50d48bf60b229e212/src/walletConnect/saga.ts).
Also we're planning to remove storing duplicated data in Redux about WC sessions. Using only the client as the single source of truth.

Now about the logs, I had a good look at them. I haven't found something indicating the true cause of the issue. But maybe you'll be able to find some signal with this.

Here are the key things that happened from the logs containing the error.
Note that the app was updated through different versions of the `@walletconnect` libs and the logs cover several days of activity. I've highlighted that too.

```
// App version running 1.62.0 (Build 1689897383 -> https://github.com/valora-inc/wallet/commit/bb38b288adcf2193540dd50b4c5b835e877cacfb, @walletconnect/sign-client 2.3.1)
INFO  [2023-07-24T07:24:12.606Z] 'NavigatorWrapper', 'Current version: 1.62.0. Required min version: 1.3.0'

// First mention of the topic 9a05b[...] that later on is shown in the error log
DEBUG [2023-07-24T08:13:17.183Z] { context: 'core' }, { context: 'core/relayer' }, { context: 'core/relayer/subscription' }, { type: 'event',
  event: 'subscription_created',
  data: 
   { topic: '9a05b6099c6cc662ed39dd43795adec83ad64dfc6cfd7779ae1bc3d6470bb5c6',
     relay: { protocol: 'irn' },
     id: '9811f5aba9ef10fe88275cfc98e997a105080997443a6a0a5d9c9c4f17f7b549' } }
     
// Session delete with the topic 9a05b[...]
DEBUG [2023-07-24T08:52:46.798Z] { context: 'core' }, { context: 'core/history' }, { type: 'event',
  event: 'history_created',
  record: 
   { id: 1690188766779475,
     topic: '9a05b6099c6cc662ed39dd43795adec83ad64dfc6cfd7779ae1bc3d6470bb5c6',
     request: 
      { method: 'wc_sessionDelete',
        params: { message: 'User disconnected.', code: 6000 } },
     chainId: undefined } }

// This warning is logged, doesn't look good.
WARN  [2023-07-24T08:53:28.966Z] { [MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 relayer_transport_closed listeners added. Use emitter.setMaxListeners() to increase limit]

// Soon after this error shows up
ERROR [2023-07-24T08:55:49.316Z] { context: 'core' }, { context: 'core/expirer' }, 'No matching key. expirer: topic:9a05b6099c6cc662ed39dd43795adec83ad64dfc6cfd7779ae1bc3d6470bb5c6'

// App version running 1.63.0 (Build 1690229229 -> https://github.com/valora-inc/wallet/commit/3f61320048660d3b028c0a9dfc0ff9b91ae468d0 with switch from @walletconnect/sign-client to @walletconnect/web3wallet 1.8.6, https://github.com/valora-inc/wallet/commit/beefa4946#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519)
INFO  [2023-07-25T06:47:27.344Z] 'NavigatorWrapper', 'Current version: 1.63.0. Required min version: 1.3.0'

// Last action request received
INFO  [2023-07-25T19:02:37.814Z] 'ValoraAnalytics', 'Tracking event wc_request_propose with properties:', { timestamp: 1690311757741,

// App version running 1.63.0 (Build 1690340360 -> https://github.com/valora-inc/wallet/commit/bf80198137745bf6db4fcbf4c0728d77d68a9f1c with upgraded from WC 2.9.0 to 2.9.1 and @walletconnect/web3wallet 1.8.7, https://github.com/valora-inc/wallet/commit/3a48f8fb8) 
INFO  [2023-07-26T07:42:50.128Z] 'NavigatorWrapper', 'Current version: 1.63.0. Required min version: 1.3.0'

// First mention of the error with 9a05b[...]
ERROR [2023-07-26T07:42:52.939Z] { context: 'client' }, [Error: No matching key. session: 9a05b6099c6cc662ed39dd43795adec83ad64dfc6cfd7779ae1bc3d6470bb5c6]

// App version running 1.63.0 (Build 1690418148 -> https://github.com/valora-inc/wallet/commit/0c04f6beb7fffa84988839d15bb078e3bb7d44c4)
INFO  [2023-07-27T07:49:03.207Z] 'NavigatorWrapper', 'Current version: 1.63.0. Required min version: 1.3.0'

// Session to curve successful
INFO  [2023-07-27T07:57:19.974Z] 'ValoraAnalytics', 'Tracking event wc_session_approve_success with properties:', { timestamp: 1690444639809,

// A sign action is attempted on curve, but no actual request received from WC, and the error is logged
DEBUG [2023-07-27T07:57:39.968Z] 'app/saga', 'Handling deep link', 'celo://wallet/wc?requestId=1690444658823705&sessionTopic=37ad78175f62a7fdfde16051c20319a1026485a1f7705b4610297ee965cbe86a'
ERROR [2023-07-27T07:57:39.983Z] { context: 'client' }, 'No matching key. session: 9a05b6099c6cc662ed39dd43795adec83ad64dfc6cfd7779ae1bc3d6470bb5c6'
ERROR [2023-07-27T07:57:39.983Z] { context: 'client' }, [Error: No matching key. session: 9a05b6099c6cc662ed39dd43795adec83ad64dfc6cfd7779ae1bc3d6470bb5c6]

// Subsequent session pairing requests work, but no more actual session requests received, with actions like signing, etc (here for Uniswap)
INFO  [2023-07-27T08:14:25.387Z] 'ValoraAnalytics', 'Tracking event wc_session_propose with properties:', { timestamp: 1690445665371,
```

Here are the full logs with all the details:
[2023-Jul.txt](https://github.com/WalletConnect/walletconnect-monorepo/files/12232132/2023-Jul.txt)


From these logs, I initially thought the upgrade from WC libs 2.9.0.to 2.9.1 may have caused the issue. But after following the same upgrades again, I wasn't able to reproduce the issue.

Do you see anything useful from these logs?

**Reply by MuckT** [reply]
@ganchoradkov I was able to pull some debug logs to highlighting the issue. It looks when we go to extend the session the `core/expirer` is also emitting `expirer_created` and is removing the session prior to being updated or acted upon by the client. The code with the session extension is in https://github.com/valora-inc/wallet/pull/4051. Thanks for taking a look!

```
 DEBUG  [05:57:55.521Z] WalletConnect/saga@handleInitialisePairing extending session start
 DEBUG  [05:57:55.523Z] {"context": "core"} {"context": "core/history"} Setting JSON-RPC request history record
 INFO   [05:57:55.524Z] {"context": "core"} {"context": "core/history"} Emitting history_created
 DEBUG  [05:57:55.524Z] {"context": "core"} {"context": "core/history"} {"event": "history_created", "record": {"chainId": undefined, "expiry": 1694152675, "id": 1691560675522109, "request": {"method": "wc_sessionExtend", "params": [Object]}, "topic": "fda5f16df318f8f7e797e9dec4963e1fa166ab1b6f91a9a7ae1ee286a9e27f42"}, "type": "event"}
 DEBUG  [05:57:55.530Z] {"context": "core"} {"context": "core/relayer"} {"context": "core/relayer/publisher"} Publishing Payload
 DEBUG  [05:57:55.530Z] {"context": "core"} {"context": "core/relayer"} {"context": "core/relayer/publisher"} Outgoing Relay Payload
 DEBUG  [05:57:55.531Z] {"context": "core"} {"context": "core/relayer"} Publishing Request Payload
 INFO   [05:57:56.095Z] {"context": "core"} {"context": "core/expirer"} Emitting expirer_created
 DEBUG  [05:57:56.096Z] {"context": "core"} {"context": "core/expirer"} {"data": {"expiration": {"expiry": 1692165475, "target": "topic:fda5f16df318f8f7e797e9dec4963e1fa166ab1b6f91a9a7ae1ee286a9e27f42"}, "target": "topic:fda5f16df318f8f7e797e9dec4963e1fa166ab1b6f91a9a7ae1ee286a9e27f42"}, "event": "expirer_created", "type": "event"}
 DEBUG  [05:57:56.604Z] {"context": "core"} {"context": "core/relayer"} Incoming Relay Payload
 DEBUG  [05:57:56.604Z] {"context": "core"} {"context": "core/relayer"} Emitting Relayer Payload
 DEBUG  [05:57:56.608Z] {"context": "core"} {"context": "core/history"} Updating JSON-RPC response history record
 DEBUG  [05:57:56.609Z] {"context": "core"} {"context": "core/history"} Updating JSON-RPC response history record
 INFO   [05:57:56.609Z] {"context": "core"} {"context": "core/history"} Emitting history_updated
 DEBUG  [05:57:56.610Z] {"context": "core"} {"context": "core/history"} {"event": "history_updated", "record": {"chainId": undefined, "expiry": 1694152675, "id": 1691560675522109, "request": {"method": "wc_sessionExtend", "params": [Object]}, "response": {"result": true}, "topic": "fda5f16df318f8f7e797e9dec4963e1fa166ab1b6f91a9a7ae1ee286a9e27f42"}, "type": "event"}
 DEBUG  [05:57:56.615Z] {"context": "core"} {"context": "core/history"} Getting record
 DEBUG  [05:57:56.616Z] {"context": "core"} {"context": "core/history"} Getting record
 DEBUG  [05:57:56.617Z] {"context": "core"} {"context": "core/history"} Deleting record
 INFO   [05:57:56.617Z] {"context": "core"} {"context": "core/history"} Emitting history_deleted
 DEBUG  [05:57:56.618Z] {"context": "core"} {"context": "core/history"} {"event": "history_deleted", "record": {"chainId": undefined, "expiry": 1694152675, "id": 1691560675522109, "request": {"method": "wc_sessionExtend", "params": [Object]}, "response": {"result": true}, "topic": "fda5f16df318f8f7e797e9dec4963e1fa166ab1b6f91a9a7ae1ee286a9e27f42"}, "type": "event"}
 DEBUG  [05:57:56.626Z] WalletConnect/saga@handleInitialisePairing extending session end
 DEBUG  [05:57:56.990Z] {"context": "core"} {"context": "core/relayer"} Incoming Relay Payload
 DEBUG  [05:57:56.990Z] {"context": "core"} {"context": "core/relayer"} Emitting Relayer Payload
 DEBUG  [05:57:56.994Z] {"context": "core"} {"context": "core/history"} Updating JSON-RPC response history record
 DEBUG  [05:57:56.995Z] {"context": "core"} {"context": "core/history"} Updating JSON-RPC response history record
 DEBUG  [05:57:56.995Z] {"context": "core"} {"context": "core/history"} Getting record
 DEBUG  [05:57:56.996Z] {"context": "core"} {"context": "core/history"} Getting record
 ERROR  [05:57:56.996Z] {"context": "client"} [Error: No matching key. history: 1691560675522109]
 ```

**Reply by ganchoradkov** [reply]
Hey @jeanregisser, I scanned through the full log and my findings are
1.  `2023-07-24T08:13:17.183Z` - the client subscribes to `9a05b6099c6cc662ed39dd43795adec83ad64dfc6cfd7779ae1bc3d6470bb5c6 ` - this seems to be the session creation
2. `2023-07-24T08:13:17.191Z` - Expirer is created for `9a05b6099c6cc662ed39dd43795adec83ad64dfc6cfd7779ae1bc3d6470bb5c6`
3. `2023-07-24T08:15:39.552Z` Several `session_requests` seem to have been received for that session and the last being @ timestamp
4. `2023-07-24T08:16:01.751Z` The app was either closed or switched to background/foreground
5. `2023-07-24T08:16:27.614Z` SDK initiating reconnection on all topics successfully
6. `2023-07-24T08:16:27.961Z` topic `9a05b6099c6cc662ed39dd43795adec83ad64dfc6cfd7779ae1bc3d6470bb5c6` subscribed successfully
7. `2023-07-24T08:52:46.796Z` Something triggering 'ValoraAnalytics', 'Tracking event wc_session_remove_start with properties
8. `2023-07-24T08:52:46.798Z` several `session_delete` requests being sent for topic `9a05b6099c6cc662ed39dd43795adec83ad64dfc6cfd7779ae1bc3d6470bb5c6` with IDs - 1690188830537368, 1690188824774706
9. `2023-07-24T08:55:49.296Z` SDK unsubscribes from the topic `9a05b6099c6cc662ed39dd43795adec83ad64dfc6cfd7779ae1bc3d6470bb5c6` due to delete requests
10. logs indicate normal usage of the wallet
11. `2023-07-26T07:42:52.931Z` first instance of `'No matching key. session: 9a05b6099c6cc662ed39dd43795adec83ad64dfc6cfd7779ae1bc3d6470bb5c6'`

Looks like you had the session active, it was disconnected (either manually or automatically triggered) and then, the session topic was attempted to be used again resulting in the errors.

As you noted 
>Also we're planning to remove storing duplicated data in Redux about WC sessions. Using only the client as the single source of truth.

This would be the recommended course of action to prevent any state inconsistencies between client storage & your redux state. 
We recommend accessing the client wherever you need to fetch sessions/topics or any additional data every time you need it.

@MuckT regarding `Error: No matching key. history` errors, feel free to ignore them, sometimes, the relay might send a message twice resulting in the error. It shouldn't cause any disruption of the SDK as we're ironing v2 behavior out

**Reply by MuckT** [reply, marked as answer]
@ganchoradkov it appears the issue was we were attempting to pair on all WalletConnect deeplinks including actions. The fix for the issue was to check for the `symKey` in the url and only attempting to pair if present: https://github.com/valora-inc/wallet/pull/4062.

It does seem a bit strange that deeplinks without symKeys would even be attempted to pair. Should [this](https://github.com/WalletConnect/walletconnect-monorepo/blob/v2.0/packages/core/src/controllers/pairing.ts#L376-L385) validation function thrown a more explicit error for attempting to pairing uri's without a `symKey`?

**Reply by glitch-txs** [reply]
Sendind this to Gancho, glad you solved this 🙏

**Reply by MuckT** [reply]
@glitch-txs we've started seeing this error again after upgrading our dependencies to `2.10.2`: [2.10.2 diff from 2.10.1](https://github.com/WalletConnect/walletconnect-monorepo/compare/2.10.1..2.10.2). Looks like it might be something to do with the cleanup of inactive pairings as a test for removing inactive pairings in `providers/universal-provider/test/index.spec.ts` was disabled in 2.10.2.

Attempting to match a request topic to active session: https://github.com/valora-inc/wallet/blob/main/src/walletConnect/saga.ts#L475-L490.

**Reply by ganchoradkov** [reply]
hey @MuckT, could you please confirm that the client instance holding the session is not overwritten in your implementation somewhow?
When the SDK receives a session request, there are several steps before emitting the event confirming the client has all needed data such as `keychain` to decrypt the data & and is actually accesing the whole session to get peer data https://github.com/WalletConnect/walletconnect-monorepo/blob/v2.0/packages/sign-client/src/controllers/engine.ts#L983

if any of the steps fail e.g. data is missing, you wouldn't receive a `session_request` event. This leads me to suspect that the client you're searching for the topic is a different instance than the client receiving the request

**Reply by MuckT** [reply]
Temporarily we've been able to downgrade our dependencies, [as seen here](https://github.com/valora-inc/wallet/pull/4406/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519), which appears to resolve the issue.

**Comment by dev-blc** [comment]
# Wallet Connect Issue: ``Error: No matching key``

## Description 
An error occurs when one or many Wallet connect requests are neither accepted nor rejected, but the app is closed, quits or exited. 
This creates a pending request that produces the below error. 
Wallet Connect used for Aleo blockchain.

`Uncaught (in promise) Error: No matching key. history: 1687441832453713`

## Solution 
Even though following all solutions listed all over the internet including clearing the local storage using both Inspect element's console and `localStorage.clear()` in JS code, the issue still persisted. 

We managed to solve the above error by the following steps 
- Go to the inspect element or window of your WebApp, dApp, or Application. 
- Now navigate to `Storage` in the top navigation bar.
- Here you can find a Wallet Connect indexed DB under the dropdown `Indexed Databases`.
- Inside the folder, check if a table is named `keyvaluestorage`.
- In this table, you can see that the `"wc@2:client:0.3:request"` row is populated.
- Make sure this row is empty or delete the table using the bin logo in the top right corner.
- Refresh your App/dApp window and try using Wallect Connect now.

#### NOTE
THE ABOVE SOLUTION IS A SUGGESTION THAT WORKED FOR US WITHOUT ANY COMPLICATIONS OR RAMIFICATIONS. PROCEED AT YOUR OWN RISK.
